core/authority_discovery/query/query_impl.cpp

```diff
@@ -123,16 +123,24 @@ namespace kagome::authority_discovery {
       queue_.pop_back();
 
       common::Buffer hash = crypto::sha256(authority);
-      std::ignore = kademlia_->getValue(
-          hash, [=](outcome::result<std::vector<uint8_t>> _res) {
-            std::unique_lock lock{mutex_};
-            --active_;
-            pop();
-            auto r = add(authority, std::move(_res));
-            if (not r) {
-              SL_DEBUG(log_, "add: {}", r.error());
-            }
-          });
+      scheduler_->schedule([=, wp = weak_from_this()] {
+        if (auto self = wp.lock()) {
+          std::ignore = kademlia_->getValue(
+              hash, [=](outcome::result<std::vector<uint8_t>> res) {
```

Comment:
I'm not actually familiar with this code, but I think this particular callback is not asynchronous. Although it would be nice if this was reflected in the name or documentation of this method.